vortex-array/src/arrays/primitive/compute/sum.rs

```diff
@@ -0,0 +1,137 @@
+use arrow_buffer::BooleanBuffer;
+use itertools::Itertools;
+use num_traits::{Float, Signed, ToPrimitive, Unsigned};
+use vortex_dtype::half::f16;
+use vortex_dtype::{NativePType, PType};
+use vortex_error::{VortexExpect, VortexResult};
+use vortex_mask::AllOr;
+use vortex_scalar::Scalar;
+
+use crate::arrays::{PrimitiveArray, PrimitiveEncoding};
+use crate::compute::SumFn;
+use crate::stats::Stat;
+use crate::variants::PrimitiveArrayTrait;
+use crate::Array;
+
+impl SumFn<&PrimitiveArray> for PrimitiveEncoding {
+    fn sum(&self, array: &PrimitiveArray) -> VortexResult<Scalar> {
+        let scalar_value = match array.validity_mask()?.boolean_buffer() {
+            AllOr::All => {
+                // All-valid
+                match array.ptype() {
+                    PType::U8 => sum_unsigned(array.as_slice::<u8>()).into(),
+                    PType::U16 => sum_unsigned(array.as_slice::<u16>()).into(),
+                    PType::U32 => sum_unsigned(array.as_slice::<u32>()).into(),
+                    PType::U64 => sum_unsigned(array.as_slice::<u64>()).into(),
+                    PType::I8 => sum_signed(array.as_slice::<i8>()).into(),
+                    PType::I16 => sum_signed(array.as_slice::<i16>()).into(),
+                    PType::I32 => sum_signed(array.as_slice::<i32>()).into(),
+                    PType::I64 => sum_signed(array.as_slice::<i64>()).into(),
+                    PType::F16 => sum_float(array.as_slice::<f16>()).into(),
+                    PType::F32 => sum_float(array.as_slice::<f32>()).into(),
+                    PType::F64 => sum_float(array.as_slice::<f64>()).into(),
```

Comment:
I think this can be only two branches:
```rust
match array.ptype() {
    x if x.is_integer() => macro!(...),
    x if x.is_float() => float_macro!(...),
}
```